velox/functions/prestosql/aggregates/AverageAggregate.cpp

```diff
@@ -132,26 +135,48 @@ class AverageAggregate : public exec::Aggregate {
 
     if (decodedRaw_.isConstantMapping()) {
       if (!decodedRaw_.isNullAt(0)) {
-        const TInput value = decodedRaw_.valueAt<TInput>(0);
+        TAccumulator value(0);
+        if constexpr (
+            std::is_same_v<TInput, UnscaledShortDecimal> ||
+            std::is_same_v<TInput, UnscaledLongDecimal>) {
+          value = decodedRaw_.valueAt<TInput>(0).unscaledValue();
+        } else {
+          value = decodedRaw_.valueAt<TInput>(0);
+        }
         const auto numRows = rows.countSelected();
         updateNonNullValue(group, numRows, TAccumulator(value * numRows));
       }
     } else if (decodedRaw_.mayHaveNulls()) {
       rows.applyToSelected([&](vector_size_t i) {
         if (!decodedRaw_.isNullAt(i)) {
-          updateNonNullValue(
-              group, TAccumulator(decodedRaw_.valueAt<TInput>(i)));
+          updateNonNullValue(group, decodedRaw_.valueAt<TInput>(i));
         }
       });
     } else if (!exec::Aggregate::numNulls_ && decodedRaw_.isIdentityMapping()) {
       const TInput* data = decodedRaw_.data<TInput>();
       TAccumulator totalSum(0);
-      rows.applyToSelected([&](vector_size_t i) { totalSum += data[i]; });
+      if constexpr (std::is_same_v<TAccumulator, int128_t>) {
+        rows.applyToSelected([&](vector_size_t i) {
+          totalSum = checkedPlus<int128_t>(totalSum, data[i].unscaledValue());
+        });
+      } else {
+        rows.applyToSelected([&](vector_size_t i) { totalSum += data[i]; });
+      }
       updateNonNullValue<false>(group, rows.countSelected(), totalSum);
     } else {
       TAccumulator totalSum(0);
-      rows.applyToSelected(
-          [&](vector_size_t i) { totalSum += decodedRaw_.valueAt<TInput>(i); });
+      if constexpr (
+          std::is_same_v<TInput, UnscaledShortDecimal> ||
+          std::is_same_v<TInput, UnscaledLongDecimal>) {
+        rows.applyToSelected([&](vector_size_t i) {
+          totalSum = checkedPlus<int128_t>(
+              totalSum, decodedRaw_.valueAt<TInput>(i).unscaledValue());
+        });
+      } else {
+        rows.applyToSelected([&](vector_size_t i) {
+          totalSum += decodedRaw_.valueAt<TInput>(i);
```

Comment:
Shouldn't we be using checkedPlus here as well?

Reply:
This is for double as intermediate type and there is no chekedPlus overloading for that.